uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseKeyboardOperation.java

```diff
@@ -133,26 +145,49 @@ private double getCellScrollDeltaX(final GridWidget gridWidget) {
         return dx;
     }
 
-    private double getCellScrollDeltaY(final GridWidget gridWidget) {
+    private double getCellScrollDeltaY(final GridWidget gridWidget, final boolean isHeaderCellSelected) {
         final GridData gridModel = gridWidget.getModel();
-        final GridData.SelectedCell origin = gridModel.getSelectedCellsOrigin();
+        final GridData.SelectedCell origin = getSelectedCellOrigin(gridModel, isHeaderCellSelected);
         final int uiRowIndex = origin.getRowIndex();
 
         double dy = 0;
         final Bounds bounds = gridLayer.getVisibleBounds();
-        final double rowHeight = gridModel.getRow(uiRowIndex).getHeight();
+        final int headerRowCount = gridModel.getHeaderRowCount();
         final double headerHeight = gridWidget.getRenderer().getHeaderHeight();
-        final double gridCellY = gridWidget.getY() + headerHeight + gridWidget.getRendererHelper().getRowOffset(uiRowIndex);
+        final double rowHeight = isHeaderCellSelected ? gridWidget.getRenderer().getHeaderRowHeight() : gridModel.getRow(uiRowIndex).getHeight();
+        final double headerYOffset = isHeaderCellSelected ? headerHeight - headerRowCount * rowHeight : headerHeight;
+        final double rowOffset = isHeaderCellSelected ? rowHeight * uiRowIndex : gridWidget.getRendererHelper().getRowOffset(uiRowIndex);
+        final double gridCellY = gridWidget.getY() + headerYOffset + rowOffset;
 
         if (gridCellY + rowHeight >= bounds.getY() + bounds.getHeight()) {
             dy = bounds.getY() + bounds.getHeight() - gridCellY - rowHeight;
-        } else if (gridCellY <= bounds.getY() + headerHeight) {
-            dy = bounds.getY() + headerHeight - gridCellY;
+        } else if (gridCellY <= bounds.getY() + headerYOffset) {
+            dy = bounds.getY() + headerYOffset - gridCellY;
         }
 
         return dy;
     }
 
+    /**
+     * It retrieves the selected cell in <code>GridData</code> model, which could be a <b>header</b> cell or a
+     * simple one.
+     * @param gridModel
+     * @param isHeaderCellSelected
+     * @return
+     */
+    private GridData.SelectedCell getSelectedCellOrigin(GridData gridModel, final boolean isHeaderCellSelected) {
+        if (isHeaderCellSelected) {
+            List<GridData.SelectedCell> selectedHeaderCells = gridModel.getSelectedHeaderCells();
+            if (KEY_RIGHT == getKeyCode()) {
+                return selectedHeaderCells.get(selectedHeaderCells.size() - 1);
+            } else {
+                return selectedHeaderCells.get(0);
+            }
+        } else {
+            return gridModel.getSelectedCellsOrigin();
+        }
+    }
```

Comment:
I do not understand this, will user be able to select just first or last header cell? @yesamer could you please clarify?

Reply:
@jomarko JAVADOC added :)

Reply:
Thank you!